packages/xsnap/test/xsnap-eof.test.js

```diff
@@ -0,0 +1,268 @@
+  return { vat, xsnapProcess };
+}
+
+test('xsnap-worker complains while trying to READ an answer when pipes are closed', async t => {
+  let xsnapProcess;
+  let toXsnap;
+  let fromXsnap;
+  let issueCommandError;
+  let vatCloseError;
+  let vatExitCode;
+  let vatExitSignal;
+  let strErr = '';
+
+  async function handleCommand(_message) {
+    // suspend the child process so the pipes appear
+    // to close atomically from its perspective
+    xsnapProcess.kill('SIGSTOP');
+    // wait for signal delivery
+    await waitFor(() => {
+      return xsnapProcess.killed;
+    });
+
+    // close the pipes
+    toXsnap.end();
+    toXsnap.destroy();
+    fromXsnap.end();
+    fromXsnap.destroy();
+
+    // un-suspend the child process
+    xsnapProcess.kill('SIGCONT');
+    // wait for signal delivery
+    await waitFor(() => {
+      return xsnapProcess.killed;
+    });
+
+    return new Uint8Array();
+  }
+
+  const setup = await doSetup(handleCommand);
+  const { vat } = setup;
+  ({ xsnapProcess } = setup);
+
+  toXsnap = xsnapProcess.stdio[3];
+  fromXsnap = xsnapProcess.stdio[4];
+
+  xsnapProcess.on('exit', (code, signal) => {
+    vatExitCode = code;
+    vatExitSignal = signal;
+  });
+
+  xsnapProcess.stdio[2].on('data', data => {
+    strErr = `${strErr}${data}`;
+  });
+
+  try {
+    // this will error out since we are intentionally
+    // breaking communications with the worker
+    await vat.issueStringCommand('0');
+  } catch (err) {
+    issueCommandError = err;
+  }
+
+  // wait for the child to exit
+  await waitFor(() => {
+    return vatExitCode || vatExitSignal;
+  });
+
+  try {
+    // this will error out since vat has died.
+    await vat.close();
+  } catch (err) {
+    vatCloseError = err;
+  }
+  t.truthy(issueCommandError, 'issueCommand() did not error out');
+  t.truthy(vatCloseError, 'vat.close() did not error out');
+  t.falsy(vatExitSignal, 'wrong vat exit signal');
+  t.is(vatExitCode, 2 /* E_IO_ERROR */, 'wrong vat exit code');
+  t.is(
+    strErr,
+    'Got EOF on netstring read. Has parent died?\n',
+    'wrong vat stderr message',
+  );
+});
+
+test('xsnap-worker complains while trying to WRITE when pipes are closed', async t => {
+  let xsnapProcess;
+  let toXsnap;
+  let fromXsnap;
+  let issueCommandError;
+  let vatCloseError;
+  let vatExitCode;
+  let vatExitSignal;
+  let strErr = '';
+
+  async function handleCommand(_message) {
+    // suspend the child process so the pipes appear
+    // to close atomically from its perspective
+    await xsnapProcess.kill('SIGSTOP');
+    // wait for signal delivery
+    await waitFor(() => {
+      return xsnapProcess.killed;
+    });
+    fromXsnap.end();
+    fromXsnap.destroy();
+    // write a fake but valid response into a pipe buffer,
+    // forcing the worker to move to the next state where it would
+    // attempt to write out return value from its own "handleCommand()"
+    // into a pipe which we've just closed above.
+    toXsnap.write('2:/0,\n');
```

Comment:
I wonder if it's worthwhile to clarify the construction of a valid response.
```suggestion
    // write a fake but valid response into a pipe buffer,
    // forcing the worker to move to the next state where it would
    // attempt to write out return value from its own "handleCommand()"
    // into a pipe which we've just closed above.
    await makeNetstringWriter(makeNodeWriter(toXsnap)).next([
      QUERY_RESPONSE_BUF,
      new Uint8Array(),
    ]);
```